tools/resource-generator/resource-generator.go

```diff
@@ -86,7 +86,11 @@ $1{{- end}}{{else}} []{{end}}`)
 	// However, when creating a template, we want its value to be something like "{{.InfraReplicas}}", which is not a uint8.
 	// Therefore, the value was substituted for a placeholder above (255). Replacing with the templated value now.
 	if strings.HasPrefix(*infraReplicasFlag, "{{") {
```

Comment:
Hi, welcome to the rabbit hole, let me show you around! :)
For the CR, we only use `resource-generator` to generate a template (`kubevirt-cr.yaml.in`), which will then be translated to proper yaml by another tool. It's called from `hack/generate.sh` [1]
`resource-generator` could also be used to generate the final CR, and I suspect it is used that way by external projects wanting to integrate KubeVirt, hence the `if`.
The problem with using `resource-generator` to generate templates, is that the tool uses `yaml.Marshall()`, which doesn't really make sense since a template can't be expected to be proper yaml. However that wasn't a problem before we made feature gates (and then infra replicas) configurable, since all values were strings [2], and we were using simple `"{{.Bla}}"` templated values, which happen to be strings. Now that we have more complex (and potentially empty) values, we need to use placeholders first and mess with the text after the marshalling...

[1] https://github.com/kubevirt/kubevirt/blob/09445bf84b9213bb44c150b470ec9b88fbe788dd/hack/generate.sh#L138
[2] https://github.com/kubevirt/kubevirt/blob/09445bf84b9213bb44c150b470ec9b88fbe788dd/tools/manifest-templator/manifest-templator.go#L69

Reply:
Well, indeed a rabbit hole of insanity :D
Thanks for the explanation!